### ISSUE TITLE
Unicode payload causing crash -- UnicodeEncodeError with six.py

### ISSUE THREAD

**Issue by jkiv** [invisibleroads/socketIO-client#81]
Hey all,

I'm connecting to a server that is sending back messages containing Unicode characters and as a result I run into the following error:

```
$ python go.py
Connected!
Traceback (most recent call last):
  File "go.py", line 21, in <module>
    socketIO.wait()
  File "c:\Python34\lib\site-packages\socketIO_client\__init__.py", line 232, in
 wait
    self._process_packets()
  File "c:\Python34\lib\site-packages\socketIO_client\__init__.py", line 254, in
 _process_packets
    for engineIO_packet in self._transport.recv_packet():
  File "c:\Python34\lib\site-packages\socketIO_client\transports.py", line 155,
in recv_packet
    six.b(packet_text))
  File "c:\Python34\lib\site-packages\six.py", line 597, in b
    return s.encode("latin-1")
UnicodeEncodeError: 'latin-1' codec can't encode character '\u2605' in position
93: ordinal not in range(256)
```

The problem character appears to be a ["BLACK STAR"](http://www.fileformat.info/info/unicode/char/2605/index.htm) 
- https://github.com/invisibleroads/socketIO-client/blob/master/socketIO_client/transports.py#L154-L155
- https://github.com/django/django/blob/master/django/utils/six.py#L596-L597

I'm using...
- Python 3.4.3
- socketIO-client 0.6.5
- six 1.9.0
- Windows 7
- "Git Bash" (Msysgit)

As an added curiosity: adding a `print(s)` in `six.py` before `return s.encode("latin-1")` causes it to not fail. No idea.

Any help appreciated.
##

**Comment by jkiv** [comment]
_Example server:_

    // node_server.js
    var app = require('http').createServer();
    var io = require('socket.io')(app);

    io.on('connection', function(socket) {
        socket.emit('Hello', '★ BLACK STAR')
    });

    app.listen(9000);

_Example client:_

    # python_client.py
    from socketIO_client import SocketIO, BaseNamespace

    class Channel(BaseNamespace):
        def on_connect(self):
            print('Connected!')

    socketIO = SocketIO('localhost', 9000, Channel)
    socketIO.wait()

**Comment by invisibleroads** [comment]
Much thanks for making an example server and client. I'll think about it and try it out in the morning. Not sure why six decided to use latin-1, and I think it should be decoding (not encoding) as the text should already be encoded.

**Comment by agijsberts** [comment]
Not exactly the same bug, but I also encountered problems with debug printing of packets with unicode characters in the ```_log``` method of ```_AbstractTransport``` in version 0.5.6. The exception is thrown because the log message is of type ```unicode``` (since the url is stored in unicode), while the arguments for the log message are instead of type ```str```. This throws a unicode decode error. As workaround I use the following monkey patch:
```
from socketIO_client import transports
def _log_utf8(self, level, msg, *attrs):
    transports._log.log(level, '[%s] %s' % (self._url, msg), *[a.decode('utf-8') for a in attrs])
transports._AbstractTransport._log = _log_utf8
```

I see that the code has been revised drastically in the subsequent versions, so I'm not sure whether the problem persists or not. Will bug fixes still be applied in the 0.5 branch? (obviously, some users cannot upgrade to version 0.6).

**Comment by ALenfant** [comment]
Just had the same problem today, maybe both encodings (latin1 and utf8) should be possible?

**Comment by kengz** [comment]
Solution from the PR above: simply change the `six.b` into `six.u` inside [transports.py](https://github.com/invisibleroads/socketIO-client/blob/master/socketIO_client/transports.py#L156). Using unicode encoding (six.u) is prolly more beneficial given it's the standard today.

**Comment by jonathanslenders** [comment]
I get exectly the same error here. 
For these two lines: https://github.com/invisibleroads/socketIO-client/blob/baae54667fe6fb998eeb5f42902088f334cf0158/socketIO_client/transports.py#L153:L154

I guess: something like ``packet_text = packet_text.encode('utf-8')`` should work.

**Comment by kengz** [comment]
Hacking it like this works, been tested and used. At the beginning of your code, import and override the property of the transport:

```python
from socketIO_client import SocketIO, WebsocketTransport
import sys
import os
import six
import socket
import websocket

def recv_packet_unicode(self):
    try:
        packet_text = self._connection.recv()
    except websocket.WebSocketTimeoutException as e:
        raise TimeoutError('recv timed out (%s)' % e)
    except websocket.SSLError as e:
        raise ConnectionError('recv disconnected by SSL (%s)' % e)
    except websocket.WebSocketConnectionClosedException as e:
        raise ConnectionError('recv disconnected (%s)' % e)
    except socket.error as e:
        raise ConnectionError('recv disconnected (%s)' % e)
    try:
        encoded = six.b(packet_text)
    except (UnicodeEncodeError):
        # print("six.b latin-l encoding fails, switching to six.u unicode uncoding")
        pass
    else:
        encoded = six.u(packet_text)
    engineIO_packet_type, engineIO_packet_data = parse_packet_text(encoded)
    yield engineIO_packet_type, engineIO_packet_data
    
# Set the new recv_packet_unicode method
WebsocketTransport.recv_packet = recv_packet_unicode
``

**Comment by invisibleroads** [comment]
@agijsberts I added your suggested fix in 7eb960ec821507bca36253f48f481923dded7b8c.

@kengz Thank you for the proposed fix. Your solution in #109 would support both latin-1 and utf-8, but I'm not sure how many socket.io servers actually use latin-1 encoding. I'll keep your proposed solution in mind if there is a request to support payloads encoded with latin-1.

For now I went with #135 from @RobbieClarken because it doesn't have the performance hit of the latin-1 test and works in both Python 2 and 3.

The changes are now live at
https://pypi.python.org/pypi/socketIO-client/0.5.7.2
and
https://pypi.python.org/pypi/socketIO-client/0.7.2
